_posts/2024-09-05-releasing-apps.md

```diff
@@ -0,0 +1,134 @@
+
+Our plan took the shape of a run book - a series of checkboxes which we’d run through top-to-bottom, which would take us from our two-trunk world, to trunk-based with release branches. These checkboxes were nuanced, and some more complex than others. We did as much work ahead-of-time as possible — creating pull requests, scripts of run to modify things in batches, and preparation of branch protection rules in parallel. When this run book eventually looked comprehensive, we trial-ran it in our sandbox, exposing a few blind-spots.
+
+When it came time to execute, we joined a call and paired on the run book - one person actioning the bullet points, the other taking notes. It took about half and hour to work through, done and ready to use an hour before engineers would start coming online!
+
+### Aiming for portability
+Across M&S we have more than 100 mobile applications in their own individual repositories. A lot of these have their own, manual release process, and a consistent branching structure doesn’t exist. We knew this coming in to the reimagining of our releasing process, and built it from the ground up with _portability_ in mind.
+
+Our release process can be summarized as just two manual steps:
+1. Create the release branch, and wait for stability.
+2. Deploy the release to production.
+
+Along the way we have automated processes that kick in and do the rest of the work: Commits to release branches submit builds to the stores, and a special cherry-pick label manages fixes destined for the release.
+
+The common points across _all_ mobile release then end up being:
+1. Releases are cut from the `main` branch.
+2. Commits to the `release/X` branch builds the apps, and stages them somewhere.
+3. Releases are “finalized”, promoting the latest build and sending any relevant communications.
+
+We aimed to build these actions in a platform- & store-agnostic way — what applies to Android should apply to iOS, and what applies to Google Play should apply to internal Endpoint management systems!
+
+We’ve recently reused the same code in iOS, leveraging reusable workflows to share portions of our process, and are generalizing more and more of the process to apply to all our other applications.
+
+### Conclusion
+Our first step towards weekly releases was to get the process right. Dual-trunks were a bottleneck at our scale, and things needed to change. Moving to a single trunk with familiar release branching simplified the day-to-day of our developers, and made championing a release a breeze.
+
+By focussing on minimizing the strain on developers writing, merging and shipping code, we open the door to more focus on quality. Our release processes now take single-digit minutes of manual intervention, freeing up developers time for more important things, like testing and polish.
```

Comment:
Can we back this up with numbers? Like,
- The number of releases we made every month before and after.
- Time taken for each release before and after. May be the time between the release cut and playstore roll-out.

It will help showcase the value of the new relesae process imho.

Reply:
> The number of releases we made every month before and after.

We're still on bi-weekly, this change enabled the move; but there are other pieces to the puzzle.

> Time taken for each release before and after. May be the time between the release cut and playstore roll-out.

That mostly depends on testing (biggest contributor now) and how many fixes they have to do on the release branch.

---

All this is encoded in

> freeing up developers time for more important things, like testing and polish.

Number-wise I think simply the two PRs -> 1 PR; automation (e.g. 0-approval process, trust the release champion), and the reduced cognitive load is the win.

Reply:
Given we don't reeeally have metrics that we've impacted here, I've opted to leave this mostly as-is. Reworked a few sections throughout the post to specifically call out cognitive load reduction though! The very last sentence alludes to the fact that this is just the _functional_ part of the work to increase release frequency. Hopefully there's room to write a follow up further down the track when we get to weekly!